src/hb-ot-color.cc

```diff
@@ -0,0 +1,213 @@
+    return OT::Null(OT::CPAL);
+
+  hb_ot_layout_t * layout = hb_ot_layout_from_face (face);
+  if (!layout->cpal) {
+    layout->cpal_blob = OT::Sanitizer<OT::CPAL>::sanitize (face->reference_table (HB_OT_TAG_CPAL));
+    layout->cpal = OT::Sanitizer<OT::CPAL>::lock_instance (layout->cpal_blob);
+  }
+
+  return *layout->cpal;
+}
+
+
+/**
+ * hb_ot_color_get_palette_count:
+ * @face: a font face.
+ *
+ * Returns: the number of color palettes in @face, or zero if @face has
+ * no colors.
+ *
+ * Since: 1.2.8
+ */
+unsigned int
+hb_ot_color_get_palette_count (hb_face_t *face)
+{
+  const OT::CPAL& cpal = _get_cpal(face);
+  return &cpal != &OT::Null(OT::CPAL) ? cpal.numPalettes : 0;
+}
+
+
+/**
+ * hb_ot_color_get_palette_name_id:
+ * @face: a font face.
+ * @palette: the index of the color palette whose name is being requested.
+ *
+ * Retrieves the name id of a color palette. For example, a color font can
+ * have themed palettes like "Spring", "Summer", "Fall", and "Winter".
+ *
+ * Returns: an identifier within @face's `name` table.
+ * If the requested palette has no name, or if @face has no colors,
+ * or if @palette is not between 0 and hb_ot_color_get_palette_count(),
+ * the result is zero.
+ *
+ * Since: 1.2.8
+ */
+unsigned int
+hb_ot_color_get_palette_name_id (hb_face_t *face, unsigned int palette)
+{
+  const OT::CPAL& cpal = _get_cpal(face);
+  if (unlikely (&cpal == &OT::Null(OT::CPAL) || cpal.version == 0 ||
+		palette >= cpal.numPalettes)) {
+    return 0;
+  }
+
+  const OT::CPALV1Tail& cpal1 = OT::StructAfter<OT::CPALV1Tail>(cpal);
+  if (unlikely (&cpal1 == &OT::Null(OT::CPALV1Tail) ||
+		cpal1.paletteLabel.is_null())) {
+    return 0;
+  }
+
+  const OT::USHORT* name_ids = &cpal1.paletteLabel (&cpal);
+  const OT::USHORT name_id = name_ids [palette];
+
+  // According to the OpenType CPAL specification, 0xFFFF means name-less.
+  // We map 0xFFFF to 0 because zero is far more commonly used to indicate
+  // "no value".
+  return likely (name_id != 0xffff) ? name_id : 0;
+}
+
+
+/**
+ * hb_ot_color_get_palette_flags:
+ * @face: a font face
+ * @palette: the index of the color palette whose flags are being requested
+ *
+ * Returns: the flags for the requested color palette.  If @face has no colors,
+ * or if @palette is not between 0 and hb_ot_color_get_palette_count(),
+ * the result is #HB_OT_COLOR_PALETTE_FLAG_DEFAULT.
+ *
+ * Since: 1.2.8
+ */
+hb_ot_color_palette_flags_t
+hb_ot_color_get_palette_flags (hb_face_t *face, unsigned int palette)
+{
+  const OT::CPAL& cpal = _get_cpal(face);
+  if (unlikely (&cpal == &OT::Null(OT::CPAL) || cpal.version == 0 ||
+		palette >= cpal.numPalettes)) {
+    return HB_OT_COLOR_PALETTE_FLAG_DEFAULT;
+  }
+
+  const OT::CPALV1Tail& cpal1 = OT::StructAfter<OT::CPALV1Tail>(cpal);
+  if (unlikely (&cpal1 == &OT::Null(OT::CPALV1Tail) ||
+		cpal1.paletteFlags.is_null())) {
+    return HB_OT_COLOR_PALETTE_FLAG_DEFAULT;
+  }
+
+  const OT::ULONG* flags = &cpal1.paletteFlags(&cpal);
```

Comment:
@khaledhosny and @behdad, with my current implementation, I’m not sure if the sanitizer will catch attacks where paletteFlags (which contains an offset relative to the beginning of the CPAL table) points beyond the limits of the CPAL structure. Does this line here (or the dereferencing on line 141) need to be changed? Or the sanitizer? Thanks for double-checking.